### PR TITLE
Support 'oneOf' in secondary indexes

### PR DESCRIPTION
**Pull request by annym**
## Overview

Description:

   Support secondary indexes on 'oneOf' fields

Why should this be merged: UFO request

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java [hunk 1]
```diff
@@ -862,6 +863,171 @@ public void testNestedSecondaryIndexesRepeatedFieldFollowedByNonPrimitive() thro
...
+     * (2) 'oneOf' not-nested in a repeated field
+     *
+     * Please see example_schemas.proto: ContactBook
+     *
+     * In this test, we create 3 contact books: work, friends & family
+     * A 'contact book' is composed of a list of 'contact', each contact has a 'name' and its 'contactInformation'
+     * is either an 'address' (which has a field city) or a 'phone number' (contactInformation is 'oneOf')
+     * A 'common' phone number, is a number shared across different contacts
+     *
+     * * ContactBookWork (3 contacts):
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:882]: 4 contacts?

File: runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java [hunk 2]
```diff
@@ -106,30 +126,62 @@
...
 
         for (Message repeatedMessage : repeatedMessages) {
             FieldDescriptor descriptor = repeatedMessage.getDescriptorForType().findFieldByName(fieldName);
+
             if (descriptor.getType().equals(FieldDescriptor.Type.MESSAGE)) {
-                // Replace upper level repeated message by next field, as we need to get all instances for this field
-                lastProcessedMessage = (Message) repeatedMessage.getField(descriptor);
-                if (lastNestedField) {
+                // Non-Primitive type
+                if (lastNestedField && isValidField(oneOfDescriptor, repeatedMessage, fieldName)) {
```
Comment (Lujie1996) [inline runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java:167]: What about this where `isValidField()` runs only once?
```
if (isValidField(oneOfDescriptor, repeatedMessage, fieldName)) {
    lastProcessedMessage = (Message) repeatedMessage.getField(descriptor);

    if (lastNestedField) {
    	// Case: last nested field (values to return, either primitive or non-primitives)
    	repeatedValues.add(ClassUtils.cast(lastProcessedMessage));
    } else {
     	// Case: intermediate nested field (accumulate messages in this level)
    	nextLevelRepeatedMessages.add(lastProcessedMessage);
    }
}
```

**Review by Lujie1996 — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java [hunk 3]
```diff
@@ -106,30 +126,64 @@
...
-                if (lastNestedField) {
-                    repeatedValues.add(ClassUtils.cast(lastProcessedMessage));
-                } else {
-                    nextLevelRepeatedMessages.add(lastProcessedMessage);
+                // Non-Primitive type
+                if (isValidField(oneOfDescriptor, repeatedMessage, fieldName)) {
+                    lastProcessedMessage = (Message) repeatedMessage.getField(descriptor);
+
+                    if (lastNestedField) {
+                        // Case: last nested field (values to return, either primitive or non-primitives)
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java:171]: can last field in the path be a message and not a primitive?

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java [hunk 4]
```diff
@@ -862,6 +863,171 @@ public void testNestedSecondaryIndexesRepeatedFieldFollowedByNonPrimitive() thro
         }
     }
 
+    /**
+     * Example to see how nested secondary indexes work when a secondary index is set on a 'oneOf' field.
+     * This test combines all possibilities of secondary indexes on 'oneOf' properties:
+     * (1) 'oneOf' index nested in a repeated field
+     *      (a) Index primitive type
+     *      (b) Index non-Primitive type
+     *      (c) Index repeated Primitive
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:872]: we may want to reword it to something like 'index repeated primitive inside a oneOf' to make it clearer?

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java [hunk 5]
```diff
@@ -862,6 +863,171 @@ public void testNestedSecondaryIndexesRepeatedFieldFollowedByNonPrimitive() thro
...
+     * In this test, we create 3 contact books: work, friends & family
+     * A 'contact book' is composed of a list of 'contact', each contact has a 'name' and its 'contactInformation'
+     * is either an 'address' (which has a field city) or a 'phone number' (contactInformation is 'oneOf')
+     * A 'common' phone number, is a number shared across different contacts
+     *
+     * * ContactBookWork (4 contacts):
+     *   - 1 with San Francisco Address
+     *   - 3 with Phone Number: 1 contact with common phone number (home and mobile)
+     *                          1 work contact with unique phone number
+     *                          1 contact with a common mobile number (its own home number)
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:886]: what is the difference between this and L884(common number home and mobile)?

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java [hunk 6]
```diff
@@ -1655,4 +1821,92 @@ public void queueOrderInTxnContext() throws Exception {
...
+
+    private List<ExampleSchemas.Contact> getContactsWithPhone(String commonMobile) {
+        // Phone Number common across all three ContactBooks
+        final ExampleSchemas.PhoneNumber phoneNumberCommon = ExampleSchemas.PhoneNumber.newBuilder()
+                .setHome("352-546-890")
+                .addMobile("611-9458741")
+                .build();
+
+        final ExampleSchemas.PhoneNumber phoneNumberWorkBook = ExampleSchemas.PhoneNumber.newBuilder()
+                .addMobile(commonMobile)
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:1843]: why do we not set this in the friends book?

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java [hunk 7]
```diff
@@ -862,6 +863,171 @@ public void testNestedSecondaryIndexesRepeatedFieldFollowedByNonPrimitive() thro
...
+        List<Uuid> contactBooksWithCommonNumber = Arrays.asList(contactBookWorkId, contactBookFamilyId);
+        List<Uuid> contactBooksWithCommonMobile = Arrays.asList(contactBookWorkId, contactBookFriendsId, contactBookFamilyId);
+        List<Uuid> contactBooksIPhone = Arrays.asList(contactBookFriendsId, contactBookFamilyId);
+
+
+        // Get by secondary index, retrieve number of contactBooks that have contacts with address in San Francisco
+        // only contactBookWork and contactBookFriends have contacts with address in SFO.
+        try (ManagedTxnContext readWriteTxn = shimStore.tx(someNamespace)) {
+            List<CorfuStoreEntry<Uuid, ExampleSchemas.ContactBook, ManagedMetadata>> contactBooks = readWriteTxn
+                    .getByIndex(table, "city", "San Francisco");
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:982]: dont we need to define the 'city' alias in the example protobuf to directly use it as the secondary index?

File: runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java [hunk 8]
```diff
@@ -64,37 +67,54 @@
...
-            if (nestedDescriptor == null) {
+            if (nestedFieldDescriptor == null) {
                 throw new IllegalArgumentException(String.format("Secondary key %s, invalid field %s", indexName, nestedFields[i]));
             }
 
-            boolean lastNestedField = (i == (nestedFields.length - 1));
+            OneofDescriptor oneOfDescriptor = inspectDescriptorForOneOfFields(nestedFieldDescriptor, oneOfFieldNameToDescriptor);
 
-            if (nestedDescriptor.isRepeated()) {
+            if (nestedFieldDescriptor.isRepeated()) {
```
[the remark this answers is not in the document]
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java:85]: Protobuf does that check already for us.. it won't compile if you try to define oneOf on a repeated field, so no need for us to confirm, it's not possible.

File: runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java (hunk 3 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java:171]: yes, we can index on non-primitives or primitives

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java (hunk 5 above)
Comment (annym) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:886]: This contact has a different home number, but a common mobile with another contact... rather L.884 has exactly the same phone number (home/mobiles)

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java (hunk 4 above)
Comment (annym) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:872]: that's what I tried to say on L.869... all these are within a oneOf, and then internally the index can be a primitive, non-primitive, or a repeated primitive.. let me reword to make it clear..

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java (hunk 7 above)
Comment (annym) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:982]: It is defined, L.168 of the protobuf

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java (hunk 6 above)
Comment (annym) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java:1843]: Because this is an example of a phone exclusive to a contact in the work book..

**Review by pankti-m — APPROVED** [review]

File: test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java (hunk 4 above)
Comment (annym) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java]: I changed it a bit to avoid confusion.

**Review by achinmishra — APPROVED** [review]

**Review by WenbinZhu — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java [hunk 9]
```diff
@@ -64,37 +67,54 @@
...
+                // Process repeated fields
+                // In this case iterate over each repeated entry, accumulate actual 'values' if its a primitive,
+                // accumulate 'messages' if its a non-primitive type (for further inspection)
                 upperLevelRepeatedField = true;
-                subMessage = processRepeatedField(subMessage, indexName,
-                        nestedFields[i], lastNestedField, repeatedMessages, repeatedValues);
+                subMessage = processRepeatedField(subMessage, indexName, nestedFields[i], lastNestedField,
```
[the remark this answers is not in the document]
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java:90]: Done.
